kv/dist_sender.go

```diff
@@ -326,28 +329,74 @@ func (ds *DistSender) getRangeDescriptor(key proto.Key) ([]proto.RangeDescriptor
-// error and must be checked in addition to the RPC error.
-func (ds *DistSender) sendRPC(desc *proto.RangeDescriptor, method string, args proto.Request, reply proto.Response) error {
+// sendRPC sends one or more RPCs to replicas from the supplied proto.Replica
+// slice. First, replicas which have gossiped addresses are corralled (and
+// rearranged depending on proximity and whether the request needs to go to a
+// leader) and then sent via rpc.Send, with requirement that one RPC to a
+// server must succeed. Returns an RPC error if the request could not be sent.
+// Note that the reply may contain a higher level error and must be checked in
+// addition to the RPC error.
+func (ds *DistSender) sendRPC(desc *proto.RangeDescriptor, method string,
+	args proto.Request, reply proto.Response) error {
 	if len(desc.Replicas) == 0 {
 		return util.Errorf("%s: replicas set is empty", method)
 	}
 
+	// Unless we know better, send the RPCs randomly.
+	order := rpc.OrderingPolicy(rpc.OrderRandom)
+	// Try to look up our own NodeDescriptor and sort replicas by proximity.
+	if nodeDesc, err := ds.gossip.GetInfo(gossip.MakeNodeIDKey(
+		proto.NodeID(ds.gossip.NodeID))); err == nil && ds.gossip.NodeID > 0 {
+		// Rearrange the replicas so that those replicas with long common prefix of
+		// attributes end up first. If there's no prefix, this is a no-op.
+		ownAttrs := nodeDesc.(*storage.NodeDescriptor).Attrs.Attrs
+		if proto.ReplicaSlice(desc.Replicas).
+			SortByCommonAttributePrefix(ownAttrs) > 0 {
+			// There's at least some attribute prefix, and we hope that the
+			// replicas that come early in the slice are now located close to
+			// us and hence better candidates.
+			order = rpc.OrderStable
+		}
+	} else if ds.gossip.NodeID > 0 {
+		// A normal node should always have that information set in Gossip when
+		// bootstrapped, so this is worth complaining about.
+		log.Warningf("own NodeDescriptor %d not available via Gossip", ds.gossip.NodeID)
+	}
+
+	// If this request needs to go to a leader and we know who that is, move
+	// it to the front and send requests in order.
+	if args.Header().ReadConsistency != proto.INCONSISTENT ||
+		proto.IsReadWrite(method) {
+		leader := ds.leaderCache.Lookup(proto.RaftID(desc.RaftID))
+		if leader != nil {
+			for i := range desc.Replicas {
+				if desc.Replicas[i].StoreID == leader.StoreID {
+					// We have found the leader, so move it to the front
+					// and make sure that we query the replicas in that order.
+					proto.ReplicaSlice(desc.Replicas).MoveToFront(i)
+					order = rpc.OrderStable
+					break
+				}
+			}
+		}
+	}
+
 	// Build a slice of replica addresses (if gossiped).
 	var addrs []net.Addr
 	replicaMap := map[string]*proto.Replica{}
 	for i := range desc.Replicas {
-		addr, err := storage.NodeIDToAddress(ds.gossip, desc.Replicas[i].NodeID)
+		nodeDesc, err := ds.gossip.GetInfo(gossip.MakeNodeIDKey(desc.Replicas[i].NodeID))
```

Comment:
I thought we'd created a handy method for all of this? We have to pull the same info out elsewhere, for the raft transport.